docs/DesignDocs/shared_memory.md

```diff
@@ -0,0 +1,192 @@
+The Open Enclave shared memory model
+====================================
+
+This document proposes a *shared memory model* that will improve
+interoperability across different secure hardware platforms, initially
+including SGX and TrustZone.
+
+Motivation
+----------
+
+Ideally, applications developed with the Open Enclave API should run on all
+supported platforms. But unfortunately, the platforms themselves may have
+incompatible shared memory models. For example, SGX implicit shares all
+untrusted memory with the enclave, whereas TrustZone shares untrusted memory
+explicitly through programmatic measures (by default, no untrusted memory
+is shared).
+
+This problem is partly addressed by the PR entitled [Enforcing full
+serialization in EDL](https://github.com/openenclave/openenclave/pull/2176),
+which proposes extending **edger8r** to print warnings for EDL specifications
+with parameters that cannot be fully serialized. Such specifications produce
+applications that implicitly share untrusted memory with the enclave. Aside
+from introducing potential security flaws, these applications are not portable
+to TrustZone and any future platforms that employ a similar model. By forcing
+full serialization, the need for shared memory can be mostly eliminated. But
+some applications need to pass untrusted shared memory to enclaves. Examples
+include:
+
+- Ring buffers
+- Context-switchless calls
+- Efficient transfer of large memory objects
+
+Whatever the reasons, a mechanism is needed for sharing untrusted memory with
+the enclave and that mechanism must be interoperable across secure hardware
+platforms. Further, the mechanism must be explicit, so operations are needed
```

Comment:
Worth noting that we don't yet guarantee that the mechanism is explicit for SGX enclaves.